### PR TITLE
SLE-825: Share Connected Mode setup among contributors

### PR DESCRIPTION
**Pull request by thahnen**
## Summary

This enables users to share their Connected Mode setup among contributors as well as import existing configurations in their workspace. This makes it cross-IDE compatible with the other SonarLint flavors!

## Info

The telemetry ticket will be done separately by me in the coming days. I want to get this PR merged first in order to have it ready for dogfooding on my side.
The usage of LSP4J classes (`Either`) will be changed to the SLCORE implementation once [SLCORE-663](https://sonarsource.atlassian.net/browse/SLCORE-663) is merged on the SLCORE side. I will tackle it separately to get rid of all the usage of the LSP4J classes.

## Testing

Due to a new issue with macOS 14 (Sonoma) that is connected to Eclipse SWT / RedDeer I cannot run integration tests at all anymore on my machine (neither from within Eclipse nor from the command line). Based on that, I decided to go with manual testing and introduce integration tests either once that issue is resolved or using my personal machine once this is on master, SLCORE is released, and stuff is on Maven Central (otherwise, my personal machine cannot access the artifacts from Repox).

Manual tests for sharing Connected Mode configuration files:
- [x] Simple Eclipse project bound to SonarQube
- [x] Simple Eclipse project bound to SonarCloud
- [x] Multi-module Maven project, all bound to SonarQube
- [x] Multi-module Maven project, one sub-module bound to different connections

Manual tests for importing Connected Mode configuration files:
- [x] Simple Eclipse project with SonarQube config
- [x] Simple Eclipse project with SonarCloud config
- [x] Simple Eclipse project with multiple configs
- [x] Simple Eclipse project with SonarQube (URL exists, project not)
- [x] Simple Eclipse project with SonarCloud (organization exists, project not)
- [x] Multi-module Maven project, all sharing the same SonarQube config
- [x] Multi-module Maven project, one sub-module with different config
- [x] Two Eclipse projects with the same SonarCloud config (same organization, same project)
- [x] Two Eclipse projects with different SonarCloud config (same organization, different project)
- [x] Two Eclipse projects with different SonarCloud config (different organization, different project)

[SLCORE-663]: https://sonarsource.atlassian.net/browse/SLCORE-663?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by jblievremont — APPROVED** [review]
Looks good from a functional point of view, nice job on handling the project hierarchies!

I haven't seen anything about telemetry, maybe it is scheduled in another ticket?

**Comment by thahnen** [comment]
Hey @jblievremont, thank you for the review 🙇

Yes, indeed, the telemetry will be tackled in [SLE-841](https://sonarsource.atlassian.net/browse/SLE-841) I will start working on the moment this PR is merged 😄

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/resource/ISonarLintProjectHierarchyProvider.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+   *  This tries to get the root project of the hierarchy itself inside the workspace. As an example, a Maven parent
+   *  project with sub-modules will return itself as it is the root project of the hierarchy.
+   *  Contract applies that {@link ISonarLintProjectHierarchyProvider#partOfHierarchy(ISonarLintProject)} was called
+   *  before and returned true.
+   *
+   *  @param project to get root project from
+   *  @return root project if found in workspace, null otherwise
+   */
+  @Nullable
+  default ISonarLintProject getRootProject(ISonarLintProject project) {
```
Comment (henryju) [inline org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/resource/ISonarLintProjectHierarchyProvider.java:64]: Does it work with Gradle? Are you sure there is only one root project? My fear with this kind of "API" is that it is only usable for Maven.

**Review by henryju — COMMENTED** [review]
LGTM but as I said I have a feeling that the new EP is tailored for Maven, and might not be used for anything else. Time will tell.

File: org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/resource/ISonarLintProjectHierarchyProvider.java (hunk 1 above)
Comment (thahnen) [reply org.sonarlint.eclipse.core/src/org/sonarlint/eclipse/core/resource/ISonarLintProjectHierarchyProvider.java:64]: It should work as Gradle also only features one root project. I did some first tests using Eclipse Buildship, and at some point, I will build a new bundle like m2e for it.

**Pull request merged**